language-support/scala/codegen/src/main/scala/com/digitalasset/codegen/CodeGen.scala

```diff
@@ -234,14 +236,16 @@ object CodeGen {
 
     // fold up the tree to discover the hierarchy's roots, each of which produces a file
     val (treeErrors, topFiles) = lf.HierarchicalOutput.discoverFiles(treeified, util)
-    val filePlans = topFiles.map { case (fil, trees) => \/-((None, fil, trees)) } ++ treeErrors
-      .map(-\/(_))
+    val filePlans = topFiles.map { case (fil, trees) =>
+      \/-((None, fil, trees)): FilePlan
+    } ++ treeErrors
+      .map(e => -\/(e): FilePlan)
 
     // Finally we generate the "event decoder" and "package ID source"
     val specials =
       Seq(lf.EventDecoderGen.generate(util, templateIds.keySet), lf.PackageIDsGen.generate(util))
 
-    val specialPlans = specials map { case (fp, t) => \/-((None, fp, t)) }
+    val specialPlans = specials map { case (fp, t) => \/-((None, fp, t)): FilePlan }
 
     filePlans ++ specialPlans
   }
```

Comment:
> I somewhat wonder if we should replace some/most uses of `\/` with `Either`. It doesn’t seem like we really use much of what scalaz provides here.

@cocreature It depends on whether you agree with the underlying rationale of scalaz/scalaz#1469. The best illustration locally is this function, `produceTemplateAndTypeFilesLF`.

As a rule, the numerous variables in this function live in a tapestry of different subtypes of the goal type. Your `ntdRights` lives in a "righty" type and `tmplLefts` lives in a "lefty" type, for example. It is only later that we unify the types via covariance, and this proceeds through later in the file.

So if at any point you inspect the various variables, you will see that they exist not at the type we necessarily need them to be, but some subtype that happens to express how they were constructed. We know it works together because when we _unify_, it lazily works. However, unless you're familiar with how the goal relates to the components' types, if you change something upstream, the occurrence of the failure (at unification) bears less relationship to the part that actually needs to change. (I am speaking generally; I would hope at this point that Scala codegen's upstream will not change significantly. This is simply the most elaborate use of "natural" covariant unification I am aware of related to eithers in the daml codebase.)

By contrast, consider the situation with invariant `\/`. The type ascriptions added here are sort of declarations of how each component variable needs to arrange itself to fit into the goal type. So we ascribe `specialPlans` as a declaration that its type may not merely exist in the subtype lattice; its plans must declaratively be `FilePlan`s, so to speak. If I do not do this, the resulting error is [mostly] apparent in how it points to `specialPlans`'s problematic [sub]type, so should such an upstream change occur, it is [mostly] apparent what part of the unification tree needs to change.

So it really boils down to whether you consider the additional explicitness useful scaffolding, or a distraction. Or, whether the magic of *variant unification is good magic or wicked magic. (There are additional issues related to higher-kinded types, but that isn't strictly relevant to what I've just described.)

Reply:
In this PR it feels more like noise to me. Maybe I’ve written Scala for so long that I’m starting to like subtyping :scream_cat:

Reply:
It certainly felt like noise while working on the PR 😆 

However, I also think I'm operating from the privileged position of updating unification code _that I happen to know already works_, which is a little different from trying to assemble new lattices.